### ISSUE TITLE
URLNormalizer#encodeNonRUICharacters fails on non-HTTP URL

### ISSUE THREAD

**Issue by andylowry** [Norconex/commons-lang#11]
The code makes use of `HttpURL.toURI(String)`, which seems to construct the URL `null://null:80` when the current invoked with a non-HTTP URL. At least it does that with a `file:` URL.

**Comment by essiembre** [comment]
URLNormalizer and HttpURL are for dealing with web resources. ``file:`` represent a file system resource and URL normalization is not well suited for it.   

What exactly are you trying to accomplish?

**Comment by andylowry** [comment]
Oh, that's a shame! The `URLNormalizer` documentation doesn't say that, at least not very loudly (I didn't read it carefully, but I did look it over.) It does mention that many of the operations are inspired by the URL RFC, which covers all possible URLs..

I'm working on a parser for JSON/YAML-based DSLs like [Swagger ](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md) and the follow-on [OpenAPI v3.0](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md) standard. I need to be able to manage "reference nodes" as defined in the [JSON Reference](https://tools.ietf.org/html/draft-pbryan-zyp-json-ref-03) definition. These take the form of (possiblly relative) URLs, with fragments interpreted as [JSON Pointers](https://tools.ietf.org/html/rfc6901) within the retrieved document.

My intention was to treat all sorts of URLs with schemes registered in the JVM uniformly, primarily `http[s]` and `file`. I'd also like to avoid loading the same document more than once, not only for performance reasons but also because if the same structure appears in multiple places within the overall document because of references, there should be just one shared internal representation of the structure in the parse result. URL normalization is a natural strategy to meet this goal more consistently, so meaningless variations in the manner of writing various references that resolve to the same structure won't be mistakenly rendered as different but "equals" objects.

I'm curious what it would take to remove this "web-only" assumption from URLNormalizer? I can say that in my testing so far, several normalization operations (primarily, the ones "guaranteed" not to yield a semantically different URL, plus dot-segment removal) are working just fine with file URLs, with the exception of encoding non-URI characters. That operation, conceptually, doesn't seem like it needs to be restricted to URLs for web resources. I assumed the use of `HttpURL` in the implementation was just an expedient, with the unfortunate side-effect of making it not work for some URL schemes.

If it doesn't seem like it would be a big undertaking, I might be able to do the work. I like `URLNormalizer quite a lot, and I'd hate to have to walk away from it. And I don't see that there'd be a downside to making this class more general, especially since it seems pretty close already.

Let me know...

Thanks.

**Comment by essiembre** [comment]
Before I make it a feature request (given such great argument for it), I had a look and ran a few tests, and it seems to work for me. It URI-encodes it properly when the ``file:`` scheme is used.  For example, here is a working test I just made with ``encodeNonURICharacters``:
```
before: file:///etc/some dir/my file.txt
after:  file:///etc/some%20dir/my%20file.txt
```
Can you provide me with examples not working for you?

**Comment by andylowry** [comment]
Hmm... it turns out that the `file:` urls I was using were not strictly correct, though commonly supported in browsers and other software.

I was getting my URLs in my test class from Java's `Class#getResource(String)` method, to locate a file that appeared alongside my test class. This resulted in a `file:` URL that omitted the double-slash that should have appeared prior to the auth-host for the URL. That part is often empty, hence we commonly see `file:` URLs like `file:///path/...`. The URLs created by Java looked like `file:/path/...`. So it would appear that my bug report should be filed with Oracle!

Because URLs like `file:/path...` are in fairly common use and tolerated by lots of software, it still might make sense for your normalizer recognize and deal with that case, but I certainly won't (now) call the current functionality a bug, and I'll add logic in my own code to add the slashes when needed.

Thanks!

**Comment by essiembre** [comment]
I modified HttpURL to relax the logic for identifying the URL protocol.  Detection is now closer to [RFC 2396](http://www.faqs.org/rfcs/rfc2396.html) syntax.

As a result, your ``file:/blah`` URLs should now be supported.  You can find that change in the latest [snapshot release](https://oss.sonatype.org/service/local/repositories/snapshots/content/com/norconex/commons/norconex-commons-lang/1.15.0-SNAPSHOT/norconex-commons-lang-1.15.0-20180603.234958-2.jar).

Please confirm.

**Comment by essiembre** [comment]
That fix is now in new [1.15.0 stable release ](https://www.norconex.com/product/commons-lang/).  Feel free to re-open if not working as expected.

**Comment by andylowry** [comment]
Thanks @essiembre. I just tested with the snapshot, and it works great!
